netconf/man/netconfd.1
.\" Process this file with
.\" nroff -e -mandoc foo.1
.\"
.TH netconfd 1 "March 23, 2025" Linux "netconfd 2.15"
.SH NAME
netconfd \- YANG-based NETCONF-over-SSH server

.SH SYNOPSIS
.nf

   netconfd [parameter=value...]

   netconfd --help [brief | normal | full]

   netconfd --version

.fi
.SH DESCRIPTION
.B netconfd
is a YANG-based NETCONF server, which can be used with
an SSH server such as OpenSSH.
This version of netconfd supports the YANG data modeling language
defined in \fBRFC 6020\fP.
.SH USAGE
Parameters can be entered in any order, and have the form:

   \fB[start] name [separator [value]]\fP

where:

    \fBstart\fP == 0, 1, or 2 dashes (foo, -foo, --foo)

    \fBname\fP == parameter name
.nf

         Parameter name completion will be attempted 
         if a partial name is entered.

.fi
    \fBseparator\fP == whitespace or equals sign (foo=bar, foo bar)

    \fBvalue\fP == string value for the parameter.
.nf

         Strings with whitespace need to be double quoted 
         (--foo="some string")

.fi
Some examples of valid command line parameters:
.nf

   foo=3
   -foo=3
   --foo=3
   foo 3
   foo=fred
   --foo "fred flintstone"
.fi

Partial parameter names can be entered if they are unique.

.SH OPTIONS
.IP --\fBaccess-control\fP=enum
Controls how the ietf-netconf-acm access control
model will be enforced during server operation.
.nf
 Enum values:
    enforcing:
      All configured access control rules will be
      enforced.
    permissive:
      All configured access control rules will be
      enforced for write and execute requests.
      All read requests will be allowed, unless
      the requested object contains the
      'nacm:very-secure' extension.  In that case,
      all configured access control rules will
      be enforced.
   disabled:
      All read, write, and execute requests will be
      allowed, unless the object contains the
      'nacm:secure' or 'nacm:very-secure' extension.
      If the 'nacm:secure' extension is in effect,
      then all configured access control rules
      will be enforced for write and execute requests.
      If the 'nacm:very-secure' extension is in effect,
      then all configured access control rules
      will be enforced for all requests.
      Use this mode with caution.
   off:
      All access control enforcement is disabled.
      Use this mode with extreme caution.
.fi
.IP --\fBaudit-log\fP=filespec
Filespec for the server audit log file to use in addition
to the normal log file or STDOUT.
.IP --\fBaudit-log-append\fP
If present, the audit log will be appended not over-written.
If not, the audit log will be over-written.
Only meaningful if the 'audit-log' parameter is also present.
.IP --\fBconfig\fP=filespec
The name of the configuration file to use.
Any parameter except this one can be set in the config file.
The default config file 
.I /etc/yuma/netconfd.conf
will not be checked if this parameter is present.
.IP --\fBdatapath\fP=list
Internal file search path for configuration data files.
Overrides the YUMA_DATAPATH environment variable.
This parameter affects the search for the startup 
configuration file (default: startup-cfg.xml).
.IP --\fBdefault-style\fP=enum 
Selects the type of filtering behavior the server will
advertise as the 'basic' behavior in the 'with-defaults'
capability.  The server will use this default handling
behavior if the 'with-defaults' parameter is not 
explicitly set.

Also, when saving a configuration to NV-storage,
this value will be used for filtering defaults
from the saved configuration.
.nf
  Enum values:
     report-all: report all values
     trim: remove leafs containing the YANG
        default value
     explicit: report only the nodes that have
        been created by the client or the server.
        This is the default value.
.fi
.IP --\fBdelete-empty-npcontainers\fP=boolean
Selects whether the server will keep or delete empty
non-presence containers in the running and startup 
configurations. Set to true to delete these containers,
and false to keep them.  Default: false.
This parameter is deprecated!  It is ignored by the server!
.IP --\fBdeviation\fP=string
 This parameter identifies a YANG module that
should only be checked for deviation statements
for external modules.  These will be collected
and applied to the real module(s) being processed.
       
Deviations are applied as patches to the target module.
Since they are not identified in the target module at
all (ala imports), they have to be specified
explicitly, so they will be correctly processed.
Zero or more instances of this parameter are allowed.
.IP --\fBeventlog-size\fP=number
Specifies the maximum number of notification events
that will be saved in the notification replay buffer.
The oldest entries will be deleted first.
The default value is  1000.
.IP --\fBfeature-disable\fP=module:feature
Identifies a feature which should be considered disabled.
Zero or more entries are allowed.
.IP --\fBfeature-enable-default\fP=boolean
If true (the default), then features will be enabled by default.
If false, then features will be disabled by default.
.IP --\fBfeature-enable\fP=module:feature
Identifies a feature which should be considered enabled.
Zero or more entries are allowed.
.IP --\fBhello-timeout\fP=number
Specifies the number of seconds that a session
may exist before the hello PDU is received.
A session will be dropped if no hello PDU
is received before this number of seconds elapses.

If this parameter is set to zero, then the server
will wait forever for a hello message, and not
drop any sessions stuck in 'hello-wait' state.

Setting this parameter to zero may permit
denial of service attacks, since only a limited
number of concurrent sessions are supported
by the server. (range 0 | 10 .. 3600).
The default value is 600 seconds (10 minutes).
.IP --\fBhelp\fP
Print this help text and exit.
The help-mode choice (--brief, --normal, or --full) may also be present
to control the amount of help text printed.
.IP --\fBhome\fP=dirspec
Directory specification for the home directory
to use instead of HOME.
.IP --\fBidle-timeout\fP=number
Specifies the number of seconds that a session
may remain idle without issuing any RPC requests.
A session will be dropped if it is idle for an
interval longer than this number of seconds.

Sessions that have a notification subscription
active are never dropped. 

If this parameter is set to zero, then the server
will never drop a session because it is idle.
(range 0 | 10 .. 360000).  The default value is
3600 seconds (1 hour).
.IP --\fBindent\fP=number
Number of spaces to indent (0..9) in formatted output.
The default is 2 spaces.
.IP --\fBlog\fP=filespec
Filespec for the log file to use instead of STDOUT.
If this string begins with a '~' character,
then a username is expected to follow or
a directory separator character.  If it begins
with a '$' character, then an environment variable
name is expected to follow.
.IP --\fBlog-append\fP
If present, the log will be appended not over-written.
If not, the log will be over-written.
Only meaningful if the \fBlog\fP parameter is
also present.
.IP --\fBlog-level\fP=enum
Sets the debug logging level for the program.
.IP --\fBmax-burst\fP=number
Specifies the maximum number of notifications
that should be sent to one session, within a
one second time interval.  The value 0 indicates 
that the server should not limit notification
bursts at all.  The default value is 10.
.IP --\fBmodpath\fP=list
Directory search path for YANG and YIN files.
Overrides the YUMA_MODPATH environment variable.
.IP --\fBmodule\fP=string
YANG or YIN source module name to load at startup.
The server will attempt to load the specified
module and its corresponding server instrumentation
library (SIL) .

If this string represents a filespec, 
ending with the \fB.yang\fP or \fB.yin\fP extension,
then only that file location will be checked.

If this string represents a module name, then
the module search path will be checked for
a file the \fB.yang\fP or \fB.yin\fP extension.

If this string begins with a '~' character,
then a username is expected to follow or
a directory separator character.  If it begins
with a '$' character, then an environment variable
name is expected to follow.
.nf

      ~/some/path ==> <my-home-dir>/some/path

      ~fred/some/path ==> <fred-home-dir>/some/path

      $workdir/some/path ==> <workdir-env-var>/some/path
.fi
.IP --\fBncxserver-sockname\fP=path
Overrides the default /tmp/ncxserver.sock UNIX
socket name netconfd listens on for incoming connections.
You have to add corresponding entry to /etc/ssh/sshd_config e.g.:
.nf
\&...
Port 1830
Subsystem netconf \"/usr/sbin/netconf-subsystem --ncxserver-sockname=/tmp/ncxserver1.sock\"
\&...
.fi
.IP --\fBport\fP=number
Specifies the TCP ports that the server will accept
connections from.  These ports must also be configured
in the /etc/ssh/sshd_config file for the SSH master
server to accept the connection and invoke the netconf
subsystem.

Up to 4 port numbers can be configured.

If any ports are configured, then only those values
will be accepted by the server.

If no ports are configured, then the server will accept
connections on the netconf-ssh port (tcp/830).
.IP --\fBprotocols\fP=bits
Specifies which NETCONF protocol versions the server
will attempt to use. The empty set is not allowed.
The values 'netconf1.0' and 'netconf1.1' are supported.
The default is to enable both NETCONF protocol versions.
.IP --\fBrunpath\fP=pathlist
Internal file search path for executable modules.
Overrides the YUMA_RUNPATH environment variable.
.IP --\fBrunning-error\fP=enum
If 'stop', then errors in the running configuration will be
treated as fatal errors.  If 'continue', the server will attempt
to continue if any validataion errors are found in the
running configuration at startup.  The default is 'stop'.
.IP --\fBstartup\fP=filespec
The full or relative filespec of the startup config file to use.
If present, overrides the default startup config
file name 'startup-cfg.xml',  This will also
override the YUMA_DATAPATH environment variable
and the datapath CLI parameter, if the first
character is the forward slash '/', indicating
an absolute file path.  If this parameter is present,
then the --no-startup and --factory-startup parameters cannot be present.
This is the default, which will cause startup-cfg.xml to
be used if not present.
.IP --\fBno-startup\fP
If present, do not load the startup config file.
Use only factory default values instead.
Does not affect the startup.cfg file, if present.
If this parameter is present, then the --startup
or --factory-startup parameter cannot be present.
.IP --\fBfactory-startup\fP
Force the system to use the factory configuration
and delete the startup config file if it exists.
Force the NV-storage startup to
contain the factory default configuration.
If this parameter is present,
then the --no-startup and --startup parameters cannot be present.
.IP --\fBstartup-error\fP=enum
If 'stop', then any errors in the startup configuration will be
treated as fatal errors.  If 'continue', the server will attempt
to continue if any errors are found in the database loaded 
from NV-storage to running at boot-time. The default is 'stop'.
.IP --\fBsubdirs\fP=boolean
If false, the file search paths for modules, scripts, and data
files will not include sub-directories if they exist in the
specified path.
      
If true, then these file search paths will include
sub-directories, if present.  Any directory name beginning
with a dot (\fB.\fP) character, or named \fBCVS\fP, will be ignored.
This is the default mode.
.IP --\fBsuperuser\fP=string
The user name to use as the superuser account.
Any session associated with this user name 
will bypass all access control enforcement.
See ietf-netconf-acm.yang for more details.
There is no default value.
.IP --\fBsystem-sorted\fP=boolean
Indicates whether ordered-by system leaf-lists 
and lists will be kept in sorted order.
The default is true.
.IP --\fBtarget\fP=enum
Specifies the database to use as the target of edit-config
operations.
.nf
  Enum values:
    running:
      Write to the running config and support the
      :writable-running capability.
    candidate:
      Write to the candidate config and support the
      :candidate and :confirmed-commit capabilities.
.fi
.IP --\fBusexmlorder\fP
If present, then XML element order will be enforced.
Otherwise, XML element order errors will not be
generated if possible. Default is no enforcement of
strict XML order.
.IP --\fBversion\fP
Print the program version string and exit.
.IP --\fBvalidate-config-only\fP
If present, netconfd acts as command line YANG
configuration validator. Load the YANG
schema modules, validate the startup
configuration and exit without opening socket and
listening for incoming sessions.
.IP --\fBwarn-idlen\fP=number
 Control whether identifier length warnings will be
generated.  The value zero disables all identifier
length checking.  If non-zero, then a warning will
be generated if an identifier is defined which 
has a length is greater than this amount.
range: 0 | 8 .. 1023.
The default value is 64.
.IP --\fBwarn-linelen\fP=number
Control whether line length warnings will be
generated.  The value zero disables all line length
checking.  If non-zero, then a warning will
be generated if the line length is greater than
this amount.  Tab characters are counted as 8 spaces.
range: 0 | 40 .. 4095.
The default value is 72.
.IP --\fBwarn-off\fP=number
Control whether the specified warning number will be
generated and counted in the warning total for the
module being parsed.
range: 400 .. 899.
This parameter may be entered zero or more times.
.IP --\fBwith-startup\fP=boolean
If set to 'true', then the :startup capability will be 
enabled. Otherwise, the :startup capability
will not be enabled.  This capability 
makes the NV-save operation an explicit operation
instead of an automatic save.  The default value is false.
.IP --\fBwith-url\fP=boolean
If set to 'false', then the :url capability will be 
disabled. Otherwise, the :url capability
will be enabled.  This capability 
allows local files to be stored as backups on the server.
The default value is true.
.IP --\fBwith-validate\fP=boolean
If set to 'true', then the :validate capability will be 
enabled. Otherwise, the :validate capability
will not be enabled.  This capability requires
extensive memory resources.  The default value is true.
.IP --\fByuma-home\fP=string
Directory for the yuma project root to use.
If present, this directory location will
override the YUMA_HOME environment variable,
if it is present.  If a zero-length string is
entered, then the YUMA_HOME environment variable
will be ignored.
.SH INPUT FILES
YANG modules can be loaded at startup with the '--module' command,
or loaded at run-time with the 'load' operation.
.SH SEARCH PATH
When a module name is entered as input, or when a
module or submodule name is specified in an import or include
statement within the file, the following search algorithm
is used to find the file:
.nf    

  1) file is in the current directory
  2) YUMA_MODPATH environment var (or set by modpath parameter)
  3) $HOME/modules directory
  4) $YUMA_HOME/modules directory
  5) $YUMA_INSTALL/modules directory OR
     default install module location, '/usr/share/yuma/modules'

.fi
By default, the entire directory tree for all locations
(except step 1) will be searched, not just the specified
directory.  The \fBsubdirs\fP parameter can be used to
prevent sub-directories from being searched.
    
Any directory name beginning with a dot character (\fB.\fP)
will be skipped.  Also, any directory named \fBCVS\fP will
be skipped in directory searches.

.SH ERROR LOGGING
By default, warnings and errors are sent to STDOUT.
    
A log file can be specified instead with the \fBlog\fP' parameter.

Existing log files can be reused with the 'logappend'
parameter, otherwise log files are overwritten.
    
The logging level can be controlled with the \fBlog-level\fP
parameter.

The default log level is 'info'.  The
log-levels are additive:
.nf

     off:    suppress all errors (not recommended!)
             A program return code of '1' indicates some error.
     error:  print errors
     warn:   print warnings
     info:   print generally interesting trace info
     debug:  print general debugging trace info
     debug2: print verbose debugging trace info
     debug3: print very verbose debugging trace info
     debug4: print maximum debugging trace info

.fi

.SH ENVIRONMENT
The following optional environment variables can be used
to control module search behavior:

.IP \fBHOME\fP
The user's home directory  (e.g., /home/andy)
.IP \fBYUMA_HOME\fP
The root of the user's Yuma work directory
(e.g., /home/andy/swdev/netconf)
.IP \fBYUMA_INSTALL\fP
The root of the directory that yangdump
is installed on this system (default is, /usr/share/yuma)
.IP \fBYUMA_DATAPATH\fP
Colon-separated list of directories to
search for data files.
(e.g.: './workdir/data-files:/home/andy/data')
The \fBdatapath\fP parameter will override this
environment variable, if both are present.
.IP \fBYUMA_MODPATH\fP
Colon-separated list of directories to
search for modules and submodules.
(e.g.: './workdir/modules:/home/andy/test-modules')
The \fBmodpath\fP parameter will override this
environment variable, if both are present.

.SH CONFIGURATION FILES
.IP \fBnetconfd.conf\fP
YANG config file
The default is: \fB/etc/yuma/netconfd.conf\fP
    
An ASCII configuration file format is supported to
store command line parameters. 

The \fBconfig\fP parameter
is used to specify a specific config file, otherwise
the default config file will be checked.
.nf    

   - A hash mark until EOLN is treated as a comment
   - All text is case-sensitive
   - Whitespace within a line is not significant
   - Whitespace to end a line is significant/
     Unless the line starts a multi-line string,
     an escaped EOLN (backslash EOLN) is needed
     to enter a leaf on multiple lines.
   - For parameters that define lists, the key components
     are listed just after the parameter name, without
     any name,  e.g.,
    

CHANGELOG
yuma123 (2.15) UNRELEASED; urgency=medium

  * 

 -- Vladimir Vassilev <vladimir@lightside-instruments.com>  Tue, 20 May 2025 14:28:55 +0200

yuma123 (2.14) stable; urgency=medium

  * Added support for notifications defined inside list
  * Bug in cli parsing list keys containing spaces fixed
  * Added agt_fd_event_cb_register(fd, callback) function for file descriptor event interrupts
  * Fixed bug in --dump-session always reporting of received chunk size 0 in *-in.ts
  * Fixed bug in val_set_cplxval_obj function always returning NO_ERR even if validation fails

 -- Vladimir Vassilev <vladimir@lightside-instruments.com>  Sun, 23 Mar 2025 15:25:36 +0100

yuma123 (2.13) stable; urgency=medium

  * ncx
    - Added if-feature parsing support for YANG 1.1 logical expressions
    - Fixed bug in obj_next_child_deep processing of models with multiple
      levels of choice/case statements
    - Fixed bug in ncx_sprintf_num for decimal64 e.g. -0.000068623 was
      represented as 0.0000-68623
  * netconfd
    - Fixed bug causing netconfd to not handle --running-error=continue
  * yangcli
    - Fill for rpc issue solved by not serializing the virtual <input>
    - Fixed socket connection timeout and signal handlng e.g. ctrl-c now
      terminates stall connection attempt
  * autoconf
    - pkg-config is now used to find libxml2
  * python/test
    - transition to python3
  * examples
    - Added initial ietf-ip support as part of the ietf-interfaces example

 -- Vladimir Vassilev <vladimir@lightside-instruments.com>  Wed, 30 Nov 2022 17:20:00 +0100

yuma123 (2.12) stable; urgency=medium

  * Added val123_add_virtual_cb to support multiple calbacks "filling" common
    virtual container.
  * Fixed anyxml support for config=true data
  * Improved /system-state/clock/current-datetime resolution to nanoseconds and
    disabled value caching
  * Augmented set-my-session and get-my-session RPCs with cache-timeout param
    allowing cache to be disabled when cache-timeout=0
  * Fixed Xpath compare of nodeset to nodeset implementation for numeric nodesets
  * Updated IETF draft YANG modules

 -- Vladimir Vassilev <vladimir@lightside-instruments.com>  Wed, 18 Aug 2021 00:08:23 +0200

yuma123 (2.11) stable; urgency=medium

  * Improved stability and interoperability
  * Added more validation testcases
  * Preliminary NMDA support rfc8342
  * RPM build scripts (from ekinzie)
  * yangcli
    - Added --use-agent parm for use of ssh-agent authentication (from ekinzie)
  * yangdump
    - YANG tree format support rfc8340

 -- Vladimir Vassilev <vladimir@transpacket.com>  Fri, 17 Aug 2018 03:02:54 +0200

yuma123 (2.10) stable; urgency=medium

  * ncx
    - YANG 1.1 Xpath functions implementation.
    - Fixed bug in handling of deviations.
    - Fixed bug in apply_commit_deletes.
    - Fixed bug in val_make_serialized_string.
    - Refactored the base64 implementation. Fixed bugs and inconsistencies.
    - Added support for multiple obj callback registrations.
    - Fixed CDATA handling.
    - Splitted yuma-netconf.yang to yuma123-netconf.yang and yuma123-netconf-types.yang
    - Various ietf draft models updated. Erratas applied.
    - Removed circular libyumancx dependencies on libyumamgr
  * agt
    - Implementation of ietf-yang-library
    - Replaced proprietary notifications with the ietf-netconf-notifications.yang defined.
    - Moved agt_arp.c integrated module as standalone example-modules/yuma-arp
  * test
    - Added netconfd testsuite netconf/test/netconfd.
    - Added litenc client side session scripting python module.
  * yangcli
    - Implemented ietf-yang-library client side support.
    - Fixed bug caused by attempted reload of already loaded module.
    - Added yes/no container fill check promp.

 -- Vladimir Vassilev <vladimir@transpacket.com>  Fri, 29 Sep 2017 13:53:28 +0200

yuma123 (2.9) stable; urgency=medium

  * Removed IETF copyrighted MIBs to avoid the complications related to +dfsg releases
  * yangrpc API refactored for stability allowing new options passed as yangcli  parameter string
  * Added dump-session yangcli parameter

 -- Vladimir Vassilev <vladimir@transpacket.com>  Sat, 20 Aug 2016 11:46:34 +0200

yuma123 (2.8+dfsg) stable; urgency=medium

  * Removed IETF copyrighted MIBs and YANG models derived from them

 -- Vladimir Vassilev <vladimir@transpacket.com>  Sun, 07 Aug 2016 15:17:40 +0200

yuma123 (2.8) stable; urgency=medium

  * Resolved bobhash license-problem-non-free-RFC-BCP78
  * Fixed unescaped dots in man pages
  * Fixed debian lintian warning out-of-date-standards-version
  * Separated libyangrpc from the libyuma package

 -- Vladimir Vassilev <vladimir@transpacket.com>  Thu, 28 Jul 2016 23:19:24 +0200

yuma123 (2.7-1) stable; urgency=medium

  * Cleaned up debian
  * Renamed yuma libraries libncx->libyumancx, libagt->libyumaagt, libmgr->libyumamgr

 -- Vladimir Vassilev <vladimir@transpacket.com>  Thu, 21 Jul 2016 18:16:16 +0200

yuma123 (2.6-1) stable; urgency=medium

  * ncx
    - Added val_make_serialized_string to the API
  * netconfd
    - Fixed a bug caused by system time adjustments and the use of nonmonotonic time function for timer callback implementation
    - Fixed bugs both in netconf 1.0 and netconf 1.1 message parsers
    - Added readline as default cli interface library with option to build with tecla instead
    - Fixed bugs in must statement validation
  * yangcli
    - Added readline as default cli interface library with option to build with tecla instead
  * examples
    - Added standalone automake based example projects building ietf-interfaces.yang and ietf-system.yang implementations for linux
  * yangrpc
    - Added yangrpc_parse_cli API parsing yangcli style command lines based on the schema context of the connection
  * python/perl
    - Initial perl and python bindings

 --  <vladimir@transpacket.com>  Sun, 17 Jul 2016 10:10:00 +0200

yuma123 (2.5-1) stable; urgency=low

  * Implemented direct-must-augment-ex:augment extension


netconf/src/netconfd/netconfd.c
/*
 * Copyright (c) 2008 - 2012, Andy Bierman, All Rights Reserved.
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.    
 */
/*  FILE: netconfd.c

                
*********************************************************************
*                                                                   *
*                  C H A N G E   H I S T O R Y                      *
*                                                                   *
*********************************************************************

date         init     comment
----------------------------------------------------------------------
04-jun-06    abb      begun; cloned from ncxmain.c
250aug-06    abb      renamed from ncxagtd.c to netconfd.c

*********************************************************************
*                                                                   *
*                     I N C L U D E    F I L E S                    *
*                                                                   *
*********************************************************************/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#ifdef MEMORY_DEBUG
#include <mcheck.h>
#endif

#include <pthread.h>

#define _C_main 1

#include  "procdefs.h"
#include  "agt.h"
#include  "agt_ncxserver.h"
#include  "agt_util.h"
#include  "agt_cli.h"
#include  "help.h"
#include  "log.h"
#include  "ncx.h"
#include  "ncxconst.h"
#include  "ncxmod.h"
#include  "status.h"
#include  "xmlns.h"


/********************************************************************
*                                                                   *
*                       C O N S T A N T S                           *
*                                                                   *
*********************************************************************/
#define NETCONFD_MOD       (const xmlChar *)"netconfd"
#define NETCONFD_EX_MOD       (const xmlChar *)"netconfd-ex"
#define NETCONFD_CLI       (const xmlChar *)"netconfd"

#define START_MSG          "Starting netconfd...\n"

/********************************************************************
 * FUNCTION load_base_schema 
 * 
 * RETURNS:
 *     status
 *********************************************************************/
static status_t load_base_schema (void)
{
    status_t res;

    /* load in the NETCONF data types and RPC methods */
    res = ncxmod_load_module( NCXMOD_YUMA_NETCONF, NULL, NULL, NULL );
    if (res != NO_ERR) {
        return res;
    }

    /* load in the NMDA NETCONF data types and RPC methods */
    res = ncxmod_load_module( "ietf-netconf-datastores", NULL, NULL, NULL );
    if (res != NO_ERR) {
        return res;
    }

    /* load in the server boot parameter definition file */
    res = ncxmod_load_module( NCXMOD_NETCONFD_EX, NULL, NULL, NULL );
    if (res != NO_ERR) {
        return res;
    }

    return res;
}   /* load_base_schema */

#ifdef NETCONFD_DEBUG_LOAD_TEST
/********************************************************************
 * FUNCTION load_debug_test_module 
 * 
 * RETURNS:
 *     status
 *********************************************************************/
static status_t load_debug_test_module( agt_profile_t* profile )
{
#define TESTMOD             (const xmlChar *)"test"
#define TESTFEATURE1        (const xmlChar *)"feature1"
#define TESTFEATURE2        (const xmlChar *)"feature2"
#define TESTFEATURE3        (const xmlChar *)"feature3"
#define TESTFEATURE4        (const xmlChar *)"feature4"

    log_debug2("\nnetconfd: Loading Debug Test Module");

    /* Load test module */
    status_t res = ncxmod_load_module( TESTMOD, NULL, 
                                       &profile->agt_savedevQ, NULL );

    if (res != NO_ERR) {
        return res;
    } else {
        agt_enable_feature(TESTMOD, TESTFEATURE1);
        agt_disable_feature(TESTMOD, TESTFEATURE2);
        agt_enable_feature(TESTMOD, TESTFEATURE3);
        agt_disable_feature(TESTMOD, TESTFEATURE4);
    }
}
#endif

/********************************************************************
 * FUNCTION load_core_schema 
 * 
 * RETURNS:
 *     status
 *********************************************************************/
static status_t load_core_schema ( agt_profile_t *profile )
{
    log_debug2("\nnetconfd: Loading NCX Module");

    /* load in the with-defaults extension module */
    status_t res = ncxmod_load_module( NCXMOD_WITH_DEFAULTS, NULL, 
                                      &profile->agt_savedevQ, NULL );
    if (res != NO_ERR) {
        return res;
    }


#ifdef NETCONFD_DEBUG_LOAD_TEST
    res = load_debug_test_module( profile );
#endif

    return res;

}  /* load_core_schema */


/********************************************************************
 * FUNCTION cmn_init
 * 
 * 
 * 
 *********************************************************************/
static status_t cmn_init ( int argc, char *argv[], boolean *showver,
                           help_mode_t *showhelpmode, boolean *validate_config_only_mode)
{
#define BUFFLEN 256

    status_t     res;
    log_debug_t  dlevel;
    int          len;
    char         buff[BUFFLEN];
    val_value_t* val;
    val_value_t* cli_valset;

    *validate_config_only_mode=FALSE;
    /* set the default debug output level */
    dlevel = LOG_DEBUG_INFO;

    /* initialize the NCX Library first to allow NCX modules to be processed.  
     * No module can get its internal config until the NCX module parser and 
     * definition registry is up */
    len = strlen(START_MSG) + strlen(COPYRIGHT_STRING_LINE0) + strlen(COPYRIGHT_STRING_LINE1) + 2;

    if (len < BUFFLEN) {
        strcpy(buff, START_MSG);
        strcat(buff, COPYRIGHT_STRING_LINE0);
        strcat(buff, COPYRIGHT_STRING_LINE1);
    } else {
        return ERR_BUFF_OVFL;
    }

    res = ncx_init( FALSE, dlevel, TRUE, buff, argc, argv);

    if (res != NO_ERR) {
        return res;
    }

    log_debug2("\nnetconfd: Loading Netconf Server Library");

    /* at this point, modules that need to read config params can be 
     * initialized */

    /* Load the core modules (netconfd and netconf) */
    res = load_base_schema();
    if (res != NO_ERR) {
        return res;
    }

    /* Initialize the Netconf Server Library with command line and conf file 
     * parameters */
    res = agt_init1(argc, argv, showver, showhelpmode);
    if (res != NO_ERR) {
        return res;
    }

    /* check quick-exit mode */
    if (*showver || *showhelpmode != HELP_MODE_NONE) {
        return NO_ERR;
    }

    cli_valset = agt_cli_get_valset();
    val = val_find_child(cli_valset, NETCONFD_EX_MOD, "validate-config-only");
    if(val!=NULL) {
        *validate_config_only_mode=TRUE;
    }

    /* Load the core modules (netconfd and netconf) */
    res = load_core_schema(agt_get_profile());
    if (res != NO_ERR) {
        return res;
    }

    /* finish initializing server data structures */
    res = agt_init2();
    if (res != NO_ERR) {
        return res;
    }

    log_debug("\nnetconfd init OK, ready for sessions\n");

    return NO_ERR;

}  /* cmn_init */


/********************************************************************
 * FUNCTION show_server_banner
 *  
 * Show startup server string
 * 
 *********************************************************************/
static void show_server_banner (void)
{
#define BANNER_BUFFLEN 32

    xmlChar buff[BANNER_BUFFLEN];
    status_t  res;

    if (LOGINFO) {
        res = ncx_get_version(buff, BANNER_BUFFLEN);
        if (res == NO_ERR) {
            log_info("\nRunning netconfd server (%s)\n", buff);
        } else {
            log_info("\nRunning netconfd server\n");
        }
    }
    
}  /* show_server_banner */


/********************************************************************
 * FUNCTION netconfd_run
 *  
 * Startup and run the NCX server loop
 * 
 * RETURNS:
 *    status:  NO_ERR if startup OK and then run OK this will be a delayed 
 *             return code some error if server startup failed
 *             e.g., socket already in use
 *********************************************************************/
static status_t
    netconfd_run (void)
{
    status_t  res;

    show_server_banner();
    res = agt_ncxserver_run();
    if (res != NO_ERR) {
        log_error("\nncxserver failed (%s)", get_error_string(res));
    }
    return res;
    
}  /* netconfd_run */


/********************************************************************
 * FUNCTION netconfd_cleanup
 *********************************************************************/
static void netconfd_cleanup (void)
{

    if (LOGINFO) {
        log_info("\nShutting down the netconfd server\n");
    }

    /* Cleanup the Netconf Server Library */
    agt_cleanup();

    /* cleanup the NCX engine and registries */
    ncx_cleanup();

}  /* netconfd_cleanup */

/********************************************************************
 * FUNCTION show_version
 *********************************************************************/
static void show_version(void)
{
    xmlChar versionbuffer[NCX_VERSION_BUFFSIZE];

    status_t res = ncx_get_version(versionbuffer, NCX_VERSION_BUFFSIZE);
    if (res == NO_ERR) {
        log_write( "\nnetconfd version %s\n", versionbuffer );
    } else {
        SET_ERROR( res );
    }
    agt_request_shutdown(NCX_SHUT_EXIT);
}

/********************************************************************
*                                                                   *
*                       FUNCTION main                               *
*                                                                   *
*********************************************************************/
int main (int argc, char *argv[])
{
    status_t           res;
    boolean            showver = FALSE;
    boolean            done = FALSE;
    help_mode_t        showhelpmode;
    boolean            validate_config_only_mode = FALSE;

#ifdef MEMORY_DEBUG
    mtrace();
#endif

    /* this loop is used to implement the restart command the sw image is not 
     * reloaded; instead everything is cleaned up and re-initialized from 
     * scratch. If the shutdown operation (or Ctl-C exit) is used instead of 
     * restart, then the loop will only be executed once */
    while (!done) {
        res = cmn_init( argc, argv, &showver, &showhelpmode, &validate_config_only_mode);
    
        if (res != NO_ERR) {
            log_error( "\nnetconfd: init returned (%s)", 
                       get_error_string(res) );
            agt_request_shutdown(NCX_SHUT_EXIT);
        } else {
            if (showver) {
                show_version();
            } else if (showhelpmode != HELP_MODE_NONE) {
                help_program_module( NETCONFD_MOD, NETCONFD_CLI, showhelpmode );
                agt_request_shutdown(NCX_SHUT_EXIT);
            } else if (validate_config_only_mode) {
                agt_request_shutdown(NCX_SHUT_EXIT);
            } else {
                res = netconfd_run();
                if (res != NO_ERR) {
                    agt_request_shutdown(NCX_SHUT_EXIT);
                }
            }
        }

        netconfd_cleanup();
        print_error_count();

        if ( NCX_SHUT_EXIT == agt_shutdown_mode_requested() ) {
            done = TRUE;
        }
    }

    print_errors();
    print_error_count();

    if ( !log_is_open() ) {
        printf("\n");
    }

#ifdef MEMORY_DEBUG
    muntrace();
#endif

    return res;
} /* main */

/* END netconfd.c */



